Add SPI to evaluate long javascript strings from explicit shared memory buffers

kkinnunen-apple opened the pull request:
#### 8bee5a463779474ff16b190214e91f1993485b29
<pre>
Add SPI to evaluate long javascript strings from explicit shared memory buffers
<a href="https://bugs.webkit.org/show_bug.cgi?id=303747">https://bugs.webkit.org/show_bug.cgi?id=303747</a>
<a href="https://rdar.apple.com/166053768">rdar://166053768</a>

Reviewed by Alex Christensen.

The SPI allows copying data to a buffer for a string once and sending
that to multiple processes explicitly. This enables sharing of the
string data between the processes.

The SPI uses explicit copying instead of VM copying, as VM copies might
share the data that exists before or after the string in the first and
last memory page.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/JSBuffer.mm
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _evaluateJavaScriptFromBuffer:withEncoding:withSourceURL:inFrame:inContentWorld:withUserGesture:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/JSBuffer.mm:
(TEST(JSBuffer, EvaluateJavaScriptFromBuffer)):

Canonical link: <a href="https://commits.webkit.org/304155@main">https://commits.webkit.org/304155@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0053ab2d96ba610f620952e75d33ea56f2d8897c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134729 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7173 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45974 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142249 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86653 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/73e40b1e-7643-4c8d-87c1-45de5f431704) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136599 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7783 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7025 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/102967 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/70246 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137676 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Passed webkitperl tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5477 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120754 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83777 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/817009c5-6760-4cd7-a6ef-f3b869c644a1) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5323 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2930 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2838 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114553 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38901 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144944 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6849 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39482 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6920 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5741 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111661 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28328 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5153 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117034 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60742 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6897 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35218 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6698 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70478 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6934 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6807 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 reviewed (approved):
Let's try it!

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -118,6 +118,13 @@ typedef NS_ENUM(NSInteger, _WKImmediateActionType) {
+typedef NS_ENUM(NSInteger, _WKJSBufferStringEncoding) {
+    _WKJSBufferStringEncodingLatin1,
+    _WKJSBufferStringEncodingUniChar
+} WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
```
We might need a #define HAVE_JSBUFFER_STRING_ENCODING for projects to be able to use this type if it exists.

kkinnunen-apple committed:
Add SPI to evaluate long javascript strings from explicit shared memory buffers
https://bugs.webkit.org/show_bug.cgi?id=303747
rdar://166053768

Reviewed by Alex Christensen.

The SPI allows copying data to a buffer for a string once and sending
that to multiple processes explicitly. This enables sharing of the
string data between the processes.

The SPI uses explicit copying instead of VM copying, as VM copies might
share the data that exists before or after the string in the first and
last memory page.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/JSBuffer.mm
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _evaluateJavaScriptFromBuffer:withEncoding:withSourceURL:inFrame:inContentWorld:withUserGesture:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/JSBuffer.mm:
(TEST(JSBuffer, EvaluateJavaScriptFromBuffer)):

Canonical link: https://commits.webkit.org/304155@main
